main.tf

```diff
@@ -453,3 +210,402 @@ resource "aws_s3_bucket_ownership_controls" "this" {
+  count  = var.acceleration_status != null ? 1 : 0
+  bucket = aws_s3_bucket.this[0].id
+  status = var.acceleration_status
+}
+
+# TODO: This should accept both ACL and Grant, exclusively
+resource "aws_s3_bucket_acl" "this" {
+  bucket = aws_s3_bucket.this[0].id
+  acl    = var.acl != "null" ? var.acl : null
+
+  dynamic "access_control_policy" {
+    for_each = length(try(jsondecode(var.access_control_policy), var.access_control_policy)) == 0 ? [] : [try(jsondecode(var.access_control_policy), var.access_control_policy)]
+
+    content {
+      dynamic "grant" {
+        for_each = lookup(access_control_policy.value, "grants")
+
+        content {
+          grantee {
+            email_address = lookup(grant.value, "email_address", null)
+            id            = lookup(grant.value, "id", null)
+            type          = lookup(grant.value, "type")
+            uri           = lookup(grant.value, "uri", null)
+          }
+
+          permission = lookup(grant.value, "permission")
+        }
+      }
+
+      owner {
+        id           = lookup(access_control_policy.value, "owner_id", null)
+        display_name = lookup(access_control_policy.value, "owner_name", null)
+      }
+    }
+  }
+}
+
+resource "aws_s3_bucket_cors_configuration" "this" {
+  count  = length(try(jsondecode(var.cors_rule), var.cors_rule)) > 0 ? 1 : 0
+  bucket = aws_s3_bucket.this[0].id
+
+  dynamic "cors_rule" {
+    for_each = try(jsondecode(var.cors_rule), var.cors_rule)
+
+    content {
+      allowed_methods = cors_rule.value.allowed_methods
+      allowed_origins = cors_rule.value.allowed_origins
+      allowed_headers = lookup(cors_rule.value, "allowed_headers", null)
+      expose_headers  = lookup(cors_rule.value, "expose_headers", null)
+      max_age_seconds = lookup(cors_rule.value, "max_age_seconds", null)
+    }
+  }
+}
+
+
+resource "aws_s3_bucket_lifecycle_configuration" "this" {
+  count  = length(try(jsondecode(var.lifecycle_rule), var.lifecycle_rule)) > 0 ? 1 : 0
+  bucket = aws_s3_bucket.this[0].id
+
+  dynamic "rule" {
+    for_each = try(jsondecode(var.lifecycle_rule), var.lifecycle_rule)
+
+    content {
+      id     = lookup(rule.value, "id", null)
+      status = lookup(rule.value, "status", null)
+
+      dynamic "abort_incomplete_multipart_upload" {
+        for_each = lookup(rule.value, "abort_incomplete_multipart_upload_days", null) == null ? [] : [lookup(rule.value, "abort_incomplete_multipart_upload_days")]
+
+        content {
+          days_after_initiation = abort_incomplete_multipart_upload.value
+        }
+      }
+
+      # Send empty map if `filter` is an empty map or absent entirely
+      dynamic "filter" {
+        for_each = length(keys(lookup(rule.value, "filter", {}))) == 0 ? [{}] : []
+
+        content {}
+      }
+
+      dynamic "filter" {
+        for_each = length(keys(lookup(rule.value, "filter", {}))) == 0 ? [] : [lookup(rule.value, "filter")]
+
+        content {
```

Comment:
Maybe we can also add the newly added optional attributes?
```hcl
          object_size_greater_than = lookup(filter.value, "object_size_greater_than", null)
          object_size_less_than    = lookup(filter.value, "object_size_less_than", null)
          prefix                   = lookup(filter.value, "prefix", null)
```

Reply:
I'd be delighted to, though I'd like to get feedback on their usage of the `and` block for predicates. For each attribute that can conditionally end up in that `and { }` segment, the check across the different filters gets more complex. Could we hoist those to an evaluation in locals maybe? Is there a straightforward way to allow passing the desired `filters` from outside the module instead?